services/subnet_manager/src/main/java/com/futurewei/alcor/subnet/service/implement/SubnetServiceImp.java

```diff
@@ -421,4 +426,175 @@ public boolean checkIfCidrOverlap(String cidr,String projectId, String vpcId) th
 
         return false;
     }
+
+    @Override
+    public void updateSubnetHostRoutes(String subnetId, InternalRouterInfo resource) throws ResourceNotFoundException, ResourcePersistenceException, DatabasePersistenceException {
+
+        // get internal routing rule
+        InternalRouterConfiguration configuration = resource.getRouterConfiguration();
+        if (configuration == null) {
+            return;
+        }
+
+        List<InternalSubnetRoutingTable> subnetRoutingTables = configuration.getSubnetRoutingTables();
+        if (subnetRoutingTables == null) {
+            return;
+        }
+
+        List<InternalRoutingRule> routingRules = null;
+        for (InternalSubnetRoutingTable internalSubnetRoutingTable : subnetRoutingTables) {
+            String internalSubnetId = internalSubnetRoutingTable.getSubnetId();
+            List<InternalRoutingRule> internalRoutingRules = internalSubnetRoutingTable.getRoutingRules();
+            if (subnetId.equals(internalSubnetId)) {
+                routingRules = internalRoutingRules;
+                break;
+            }
+        }
+
+        // get List<HostRoute> in subnet entity
+        SubnetEntity subnetEntity = this.subnetDatabaseService.getBySubnetId(subnetId);
+        if (subnetEntity == null) {
+            return;
+        }
+        List<HostRoute> hostRoutes = subnetEntity.getHostRoutes();
+
+        if (hostRoutes == null || routingRules == null) {
+            return;
+        }
+
+        // update subnet routes
+        for (InternalRoutingRule internalRoutingRule : routingRules) {
+            String operationType = internalRoutingRule.getOperationType().getOperationType();
+            String destination = internalRoutingRule.getDestination();
+            String nextHopIp = internalRoutingRule.getNextHopIp();
+            if (destination == null || operationType == null) {
+                continue;
+            }
```

Comment:
It's not allow `destination` or `operationType` equal to `null` in the `InternalRouterInfo`. Should we catch this exception instead of just continue?

Reply:
NP